### ISSUE TITLE
CUDA 11 support

### ISSUE THREAD

**Issue by aydwi** [RavenCommunity/kawpowminer#54]
Could we have a release which supports **CUDA 11.x**.

Current version `kawpowminer 1.2.3+commit.d46045ef` fails on **CUDA 11.x** machines with something like this:

```
Unexpected error CUDA error in func set_constants at line 180 calling cudaMemcpyToSymbol(d_dag, &_dag, siz
eof(hash64_t*)) failed with error invalid device symbol on CUDA device 0a:00.0
```

**Comment by aydwi** [comment]
To those searching, the following repo works as of now on GPUs utilizing CUDA 11 if you build from source: https://github.com/one-won/kawpowminer.git

Thanks to @one-won

**Comment by chris-ayala** [comment]
I'm getting the same error on a 3090. Built @one-won's repo but unfortunately, I'm getting the same error. Did you have to do anything special beyond what the build instructions say?

**Comment by greenku** [comment]
My understanding in researching the sm_ code needed for the 30 series cards comes from https://arnon.dk/matching-sm-architectures-arch-and-gencode-for-various-nvidia-cards/ and I see Ampere (CUDA 11 and later). Quick check of the one-won repo above does not include the sm_86, compute_86 entry in kawpowminer/libethash-cuda/CMakeLists.txt

**SM80 or SM_80, compute_80** – 
NVIDIA A100 (the name “Tesla” has been dropped – GA100), NVIDIA DGX-A100
**SM86 or SM_86, compute_86** – (from CUDA 11.1 onwards)
Tesla GA10x cards, RTX Ampere – RTX 3080, GA102 – RTX 3090, RTX A6000, RTX A40, GA106 – RTX 3060, GA104 – RTX 3070, GA107 – RTX 3050

**Comment by grenelt** [comment]
> I'm getting the same error on a 3090. Built @one-won's repo but unfortunately, I'm still getting the same error. Did you have to do anything special beyond what the build instructions say?
> 
> I have cuda_11.0.2_451.48 installed. Maybe I should update?

How did you built it? I'm getting errors with w10/vs2019 (and cuda_11.2.1_461.09) - but i don't understand them:

I got the same error twice:
`CMake Error at C:/.hunter/_Base/Download/Hunter/0.23.214/e14bc15/Unpacked/cmake/modules/hunter_error_page.cmake:12 (message):`

... but with no useful message.

One of them i got rid of with an info in #29:

> Change ./cmake/Hunter/config.cmake to
> hunter_config(Boost VERSION 1.72.0-p1)
> solved this for me (Catalina 10.15.7)

**Comment by kebeckwith** [comment]
@chris-ayala @grenelt 

It seems that the RTX 30xx series actually uses compute version 8.6, I see @one-won's changes go up to compute version 8.0 which didn't work for my 3080 also.

I created a fork that includes compute version 8.6: https://github.com/kebeckwith/kawpowminer.git

The build takes a long time because I included compatibility for all compute versions that CUDA 11 supports, but the build and software work great on my 3080, using CUDA 11.2.2_461.33 for the build.

**Comment by chris-ayala** [comment]
@kebeckwith thank you for sharing your fork. I confirmed it working on my 3090 using CUDA 11.2, V11.2.142. Great work!

**Comment by greenku** [comment]
Do you have a single GPU? I also am able to get one GPU to work by changing the compute version to 86, but the second and third GPUs would still abort and not mine.

**Comment by kebeckwith** [comment]
Yeah that's correct, I only have the one. Are all of the GPUs you have RTX 3000 series, and are they in the same machine?

**Comment by greenku** [comment]
I have two 3070s and one 2060 in the same machine. The "first" GPU is one of the 3070s and it loads and mines fine. The 2060 and second 3070 error out. I also have built ethminer on this machine and it runs all three cards fine. To me, that points to kawpowminer having the issue as both ethminer and kawpowminer have been built using the same version of CUDA.

**Comment by fdoving** [comment]
Relayed message from : QxQNR8uZoO8qWt2S: 

just patched kawpowminer to work with more than 1 30XX card, below is the patch:
https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__PRIMARY__CTX.html
Releasing a primary context that has not been previously retained will fail with CUDA_ERROR_INVALID_CONTEXT. 

line 69 in CUDAMiner.cpp has to enwrapped with try { line69 } catch(...) {}

![image](https://user-images.githubusercontent.com/26197922/118684219-53c81900-b802-11eb-815b-305094f1bff1.png)

**Comment by fdoving** [comment]
Should be fixed with 4718b079756d44a48c1333e55fdeef65c7e12e99

Please test and give feedback.
